File: ecommerce/src/main/java/com/pedrooliveira/ecommerce/service/ProdutoService.java
package com.pedrooliveira.ecommerce.service;

import com.pedrooliveira.ecommerce.exception.ResourceNotFoundException;
import com.pedrooliveira.ecommerce.model.Produto;
import com.pedrooliveira.ecommerce.repository.ProdutoRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

@Service
public class ProdutoService {

    @Autowired
    private ProdutoRepository produtoRepository;

    public Produto salvar(Produto produto) {
        return produtoRepository.save(produto);
    }

    public List<Produto> listarTodos() {
        return produtoRepository.findAll();
    }

    public Produto buscarPorId(Long id) {
        return produtoRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Produto não encontrado com id: " + id));


    }

    public void deletar(Long id) {
        produtoRepository.deleteById(id);
    }

    public Produto atualizar(Long id, Produto produtoAtualizado) {
        Produto produto = buscarPorId(id);
        produto.setNome(produtoAtualizado.getNome());
        produto.setDescricao(produtoAtualizado.getDescricao());
        produto.setPreco(produtoAtualizado.getPreco());
        produto.setEstoque(produtoAtualizado.getEstoque());
        return produtoRepository.save(produto);
    }

    public Page<Produto> listarTodos(Pageable pageable) {
        return produtoRepository.findAll(pageable);
    }

    public Page<Produto> buscarPorNome(String nome, Pageable pageable) {
        return produtoRepository.findByNomeContainingIgnoreCase(nome, pageable);
    }

    public Page<Produto> buscarPorPreco(Double precoMin, Double precoMax, Pageable pageable) {
        return produtoRepository.findByPrecoBetween(precoMin, precoMax, pageable);
    }

}

File: ecommerce-frontend/src/services/produtoService.ts
import api from './api';
import { Produto, PageResponse } from '@/types';

export const produtoService = {
    async listar(page = 0, size = 10): Promise<PageResponse<Produto>> {
        const response = await api.get(`/produtos?page=${page}&size=${size}`);
        return response.data;
    },

    async buscarPorId(id: number): Promise<Produto> {
        const response = await api.get(`/produtos/${id}`);
        return response.data;
    },

    async criar(produto: Produto): Promise<Produto> {
        const response = await api.post('/produtos', produto);
        return response.data;
    },

    async atualizar(id: number, produto: Produto): Promise<Produto> {
        const response = await api.put(`/produtos/${id}`, produto);
        return response.data;
    },

    async deletar(id: number): Promise<void> {
        await api.delete(`/produtos/${id}`);
    },

    async buscarPorNome(nome: string, page = 0): Promise<PageResponse<Produto>> {
        const response = await api.get(`/produtos/buscar?nome=${nome}&page=${page}`);
        return response.data;
    },

    async buscarPorPreco(precoMin: number, precoMax: number, page = 0): Promise<PageResponse<Produto>> {
    const response = await api.get(`/produtos/preco?precoMin=${precoMin}&precoMax=${precoMax}&page=${page}`);
    return response.data;
    }
};